perception/traffic_light_recognition/traffic_light_map_based_detector/src/node.cpp

```diff
@@ -17,54 +17,68 @@
-  camera_info_sub_ =
-    pnh_.subscribe("input/camera_info", 1, &MapBasedDetector::cameraInfoCallback, this);
-  route_sub_ = pnh_.subscribe("input/route", 1, &MapBasedDetector::routeCallback, this);
-  roi_pub_ = pnh_.advertise<autoware_perception_msgs::TrafficLightRoiArray>("output/rois", 1);
-  viz_pub_ = pnh_.advertise<visualization_msgs::MarkerArray>("debug/markers", 1);
-  pnh_.getParam("max_vibration_pitch", config_.max_vibration_pitch);
-  pnh_.getParam("max_vibration_yaw", config_.max_vibration_yaw);
-  pnh_.getParam("max_vibration_height", config_.max_vibration_height);
-  pnh_.getParam("max_vibration_width", config_.max_vibration_width);
-  pnh_.getParam("max_vibration_depth", config_.max_vibration_depth);
+
+  using std::placeholders::_1;
+
+  // subscribers
+  map_sub_ = create_subscription<autoware_lanelet2_msgs::msg::MapBin>(
+    "input/vector_map", 1, std::bind(&MapBasedDetector::mapCallback, this, _1));
+  camera_info_sub_ = create_subscription<sensor_msgs::msg::CameraInfo>(
+    "input/camera_info", 1, std::bind(&MapBasedDetector::cameraInfoCallback, this, _1));
+  route_sub_ = create_subscription<autoware_planning_msgs::msg::Route>(
+    "input/route", 1, std::bind(&MapBasedDetector::routeCallback, this, _1));
+
+  // publishers
+  roi_pub_ = this->create_publisher<autoware_perception_msgs::msg::TrafficLightRoiArray>(
+    "output/rois", 1);
+  viz_pub_ = this->create_publisher<visualization_msgs::msg::MarkerArray>(
+    "debug/markers", 1);
+
+  // parameter declaration needs default values: are 0.0 goof defaults for this?
+  config_.max_vibration_pitch = declare_parameter<double>("max_vibration_pitch", 0.0);
+  config_.max_vibration_yaw = declare_parameter<double>("max_vibration_yaw", 0.0);
+  config_.max_vibration_height = declare_parameter<double>("max_vibration_height", 0.0);
+  config_.max_vibration_width = declare_parameter<double>("max_vibration_width", 0.0);
+  config_.max_vibration_depth = declare_parameter<double>("max_vibration_depth", 0.0);
 }
 
-void MapBasedDetector::cameraInfoCallback(const sensor_msgs::CameraInfo::ConstPtr & input_msg)
+  void MapBasedDetector::cameraInfoCallback(const sensor_msgs::msg::CameraInfo::ConstSharedPtr input_msg)
 {
   camera_info_ptr_ = input_msg;
   if (lanelet_map_ptr_ == nullptr || camera_info_ptr_ == nullptr) return;
 
-  autoware_perception_msgs::TrafficLightRoiArray output_msg;
+  autoware_perception_msgs::msg::TrafficLightRoiArray output_msg;
   output_msg.header = camera_info_ptr_->header;
-  geometry_msgs::PoseStamped camera_pose_stamped;
+  geometry_msgs::msg::PoseStamped camera_pose_stamped;
   try {
-    geometry_msgs::TransformStamped transform;
+    geometry_msgs::msg::TransformStamped transform;
     transform = tf_buffer_.lookupTransform(
-      "map", input_msg->header.frame_id, input_msg->header.stamp, ros::Duration(0.2));
+       "map", input_msg->header.frame_id, tf2::TimePointZero);
```

Comment:
I think we've found out that lookupTransform with timeout works if timestamp is not zero so could you change this to use the exact time with timeout?
e.g. 
```
tf2_buffer_.lookupTransform(target_frame, source_frame, time_stamp, rclcpp::Duration::from_seconds(0.2));
```

Reply:
Do we have a working example? BufferCore doesn't seem to accept such a function. Should we switch to back Buffer?

Reply:
@mitsudome-r I have pushed the above changes, excluding the lookuptransform. Let me know if you want to to revisit this.

Reply:
> I think we've found out that lookupTransform with timeout works if timestamp is not zero so could you change this to use the exact time with timeout?
@mitsudome-r  Strangely that is only the case with `waitForTransform` I think, reading the logic of `tf2_ros::lookupTransform`, it seems like it should work.

I am still a bit confused with what's the issue here so here's my take:

* If the aim is to have `lookupTransform` wait until the transform is available, then using `tf2_ros::Buffer::lookupTransform` with the message timestamp + timeout should be used.
* If the aim is to work with the current state of the buffer and not wait for anything then:
   * Use lookupTransform with exact time stamp -> You might not succeed if the data has not arrived just yet.
   * Or do something more fail-safe:

```cpp
if(tf_buffer_.can_transform(/* */, time_stamp)){
   tf = tf_buffer_. lookupTransform(/* */, time_stamp);
} else {  // Data hasn't arrived yet, use the best we got:
   tf = tf_buffer_. lookupTransform(/* */, tf2::TimePointZero);
}
```

Reply:
@cvasfi @mitsudome-r 
Thanks for the suggestions.
Used `tf2_ros::Buffer` and `lookupTransform` with message timestamp and timeout.

Something not obvious (to me), I had to instantiate the Buffer object with the node clock or else there is a compilation error in the TransformListener instantiation.